src/Illuminate/Database/Query/Grammars/Grammar.php

```diff
@@ -134,28 +134,35 @@ protected function compileJoins(Builder $query, $joins)
         foreach ($joins as $join) {
             $table = $this->wrapTable($join->table);
 
-            // First we need to build all of the "on" clauses for the join. There may be many
-            // of these clauses so we will need to iterate through each one and build them
-            // separately, then we'll join them up into a single string when we're done.
-            $clauses = [];
+            $type = $join->type;
 
-            foreach ($join->clauses as $clause) {
-                $clauses[] = $this->compileJoinConstraint($clause);
-            }
+            // First build the JOIN
+            $sql[] = $type.' join '.$table;
 
-            // Once we have constructed the clauses, we'll need to take the boolean connector
-            // off of the first clause as it obviously will not be required on that clause
-            // because it leads the rest of the clauses, thus not requiring any boolean.
-            $clauses[0] = $this->removeLeadingBoolean($clauses[0]);
+            // build the ON clauses only if provided. In custom closure for $join->one it's not
+            // necessary to provide on clause
+            if (!empty($join->clauses)) {
+                // First we need to build all of the "on" clauses for the join. There may be many
```

Comment:
Instead if nesting this and changing all the indention, use a continue statement.